### PR TITLE
ARROW-9068: [C++][Dataset] Simplify partitioning interface

### PR DESCRIPTION
**Pull request by bkietz**
Format and Parse now operate on a whole path rather than on individual segments.

This will simplify building WritePlans but I've only made minimal changes there (to keep tests passing); follow up in https://issues.apache.org/jira/browse/ARROW-8382

### CODE REVIEW

File: cpp/src/arrow/dataset/partition.h [hunk 1]
```diff
@@ -59,21 +59,10 @@ class ARROW_DS_EXPORT Partitioning {
...
-                                                    int i) const = 0;
-
-  virtual Result<std::string> Format(const Expression& expr, int i) const {
-    // FIXME(bkietz) make this pure virtual
-    return Status::NotImplemented("formatting paths from ", type_name(), " Partitioning");
-  }
-
   /// \brief Parse a path into a partition expression
-  Result<std::shared_ptr<Expression>> Parse(const std::string& path) const;
+  virtual Result<std::shared_ptr<Expression>> Parse(const std::string& segment) const = 0;
```
Comment (fsaintjacques) [inline cpp/src/arrow/dataset/partition.h:63]: Name the parameter path.

**Review by fsaintjacques — APPROVED** [review]
LGTM, one minor name fix. Should help with WritePlan simplificaiton.

File: cpp/src/arrow/dataset/partition.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/partition.h:63]:
```suggestion
  virtual Result<std::shared_ptr<Expression>> Parse(const std::string& path) const = 0;
```

**Comment by kszucs** [comment]
Thanks Francois for the review! Merging on green.

**Comment by kszucs** [comment]
Appveyor error is unrelated.

**Pull request closed without merging**